src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -67,190 +68,187 @@ async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)>
+            // First find symbols of all applicable extension methods.
+            // Workspace's syntax/symbol index is used to avoid iterating every method symbols in the solution.
+            var results = await GetExtensionMethodSymbolsAsync(
+                document,
+                position,
+                receiverTypeSymbol,
+                namespaceInScope,
+                forceIndexCreation,
+                cancellationToken).ConfigureAwait(false);
+
+            counter.GetSymbolsTicks = Environment.TickCount - ticks;
+            ticks = Environment.TickCount;
+
+            var indicesComplete = true;
+            using var _1 = PooledDictionary<INamespaceSymbol, string>.GetInstance(out var namespaceNameCache);
+            using var _2 = PooledDictionary<(string containingNamespace, string methodName, bool isGeneric), (IMethodSymbol bestSymbol, int overloadCount)>.GetInstance(out var overloadMap);
+
+            // Aggregate overloads
+            foreach (var result in results)
             {
-                if (receiverTypeSymbol is ITypeParameterSymbol typeParameter)
+                // `null` indicates we don't have the index ready for the corresponding project/PE,
+                // we will queue a background task to force creating them. Meanwhile, returns
+                // what we do have even it means we only show partial results.
+                if (result == null)
                 {
-                    foreach (var constraintType in typeParameter.ConstraintTypes)
-                    {
-                        AddNamesForTypeWorker(constraintType, builder);
-                    }
+                    indicesComplete = false;
+                    continue;
                 }
-                else
+
+                foreach (var symbol in result)
                 {
-                    builder.Add(GetReceiverTypeName(receiverTypeSymbol));
-                    builder.AddRange(receiverTypeSymbol.GetBaseTypes().Select(t => t.MetadataName));
-                    builder.AddRange(receiverTypeSymbol.GetAllInterfacesIncludingThis().Select(t => t.MetadataName));
+                    var containingNamespacename = GetFullyQualifiedNamespaceName(symbol.ContainingNamespace, namespaceNameCache);
 
-                    // interface doesn't inherit from object, but is implicitly convertible to object type.
-                    if (receiverTypeSymbol.IsInterfaceType())
-                    {
-                        builder.Add(nameof(Object));
-                    }
+                    // Select the overload with minimum number of parameters to display
+                    (var bestSymbol, var overloadCount) = overloadMap.TryGetValue((containingNamespacename, symbol.Name, symbol.Arity > 0), out var currentValue)
+                        ? (currentValue.bestSymbol.Parameters.Length > symbol.Parameters.Length ? symbol : currentValue.bestSymbol, currentValue.overloadCount + 1)
+                        : (symbol, 1);
```

Comment:
i'll be honest.  this complexity breaks me

Reply:
i think this is an example of code that is better imperative.  because thsi code is explicitly attempting to mutate things.  so trying ot jam it all into complex expressions that mutate with an assignemnt doesn't actually help my brain.